### ISSUE TITLE
enhancement: allow i3bar to not display some workspaces

### ISSUE THREAD

**Issue by woodensquares** [i3/i3#2333]
I have been using this

https://gist.github.com/woodensquares/c1afc4fb56b4d9d21fa261fb4b28b092

to allow me to have some workspaces hidden from i3bar, not sure if there is any interest in having this be generally available, if there is please let me know. 

As a side note for fun I have taken a stab also at writing a test that exercises this but I am having some issues.

The test creates a couple of workspaces, with windows in each, and I was planning to send button presses and then verify which window was displayed, which would help infer if a to-be-ignored workspace was displayed or not, but even if I have xephyr up with, say, workspace 1 + 2 and 2 focused, and use xtest_button_press/release(1, 500, 500); it will switch to workspace 1, when I would think it should stay on workspace 2 (as I am not clicking anywhere near the top-left i3bar workspace 1 pager)

Is xtest_button_press supposed to work for this kind of test? the only use I have found in the test suite is to test binding to a specific button nothing about checking behavior with mouse position.

**Comment by Airblader** [comment]
Related to #1808 and #2214.

**Comment by Airblader** [comment]
Personally I think hiding workspaces from i3bar is OK, but I don't like the approach with "hide them if their name ends a certain way". I think what @acrisci proposed, which is having underscore(s) in the beginning, is more consistent with other things (e.g., display of marks). Another approach would be a regular expression.

To judge the usefulness of the feature it'd be good if you could also explain what you are using this for, i.e., why you want to hide workspaces?

**Comment by woodensquares** [comment]
to be honest I had a hardcoded 'starts with _' for months that I have been using, but when thinking about sharing it I thought using a trailing symbol (whatever the user decided) would be better because trailing characters seem to be ignored by i3 for "num" (this way you can do scripts creating a new "hidden" workspace automatically by incrementing num of the last hidden, and appending your ignore characters)

I will write a blog post about my use of i3 in the near future, but basically for me my linux desktop is running only i3, terminals if I need something, emacs and virtualbox, and everything else I do pretty much is in virtualbox vms. I have customized i3bar status scripts that display all the VMs currently running, and have them set up so I can click on the vm name and i3 is switched to the workspace of the vm in question.

This means that having the VM workspaces displayed in i3bar is redundant and just takes space, I'd rather have all my virtualbox VMs in hidden workspaces and switch to them via my custom i3bar status handlers. I do think this use is a bit niche, but it's working well for me and I was wondering if others might be interested.

**Comment by Airblader** [comment]
Out of curiosity, why don't you just name the VM workspaces after the VM running on it and then use the workspace button?

**Comment by woodensquares** [comment]
because it looks nicer with the custom i3status I have, for example right now on my center monitor I have something like

[1r][emacs]            devel: 4% 0.4GB | web: 15% 1.5GB | services: 5% 0.6GB | firewall: 3% 0.0GB | 2016-05-01 14:33:20

where I have a terminal in 1r, emacs in [emacs], and several VMs running, the status bar displays the load of each vm, and how much memory it's using, and I can click on it and via the i3bar handler it switches to the workspace of the actual VM

I know, I know, a bit niche but it's cool IMHO

**Comment by Airblader** [comment]
OK, I see. When providing dynamic information it makes even more sense, of course. Looking forward to that blog post. :-)

**Comment by woodensquares** [comment]
hah, sure, I love that i3 is so easily extensible, allowing these kind of customizations

**Comment by orestisf1993** [comment]
I think this is good idea. The core change can be much smaller:

```diff
diff --git a/i3bar/src/workspaces.c b/i3bar/src/workspaces.c
index 7285d150..0ffc3ef0 100644
--- a/i3bar/src/workspaces.c
+++ b/i3bar/src/workspaces.c
@@ -104,6 +104,10 @@ static int workspaces_string_cb(void *params_, const unsigned char *val, size_t
 
     if (!strcmp(params->cur_key, "name")) {
         const char *ws_name = (const char *)val;
+        if (config.ignore_ws_starting_with &&
+            STARTS_WITH(ws_name, strlen(config.ignore_ws_starting_with), config.ignore_ws_starting_with)) {
+            return 1;
+        }
         params->workspaces_walk->canonical_name = sstrndup(ws_name, len);
 
         if ((config.strip_ws_numbers || config.strip_ws_name) && params->workspaces_walk->num >= 0) {
@@ -147,6 +151,11 @@ static int workspaces_string_cb(void *params_, const unsigned char *val, size_t
     }
 
     if (!strcmp(params->cur_key, "output")) {
+        if (!params->workspaces_walk->name) {
+            /* Workspace was ignored. */
+            return 1;
+        }
+
         /* We add the ws to the TAILQ of the output, it belongs to */
         char *output_name = NULL;
         sasprintf(&output_name, "%.*s", len, val);
```

Matching workspaces to hide with regex is much better but we'd need to compile i3bar with libpcre.

**Comment by Airblader** [comment]
> Matching workspaces to hide with regex is much better but we'd need to compile i3bar with libpcre.

I don't think that's an issue. i3bar comes with i3, not separately, so the dependency is de-facto already there anyway.

**Comment by orestisf1993** [comment]
With regex: https://github.com/i3/i3/compare/next...orestisf1993:i3bar-ignorews  
Without: https://github.com/i3/i3/compare/next...orestisf1993:i3bar-ignorews-regex

Which one do you prefer?

**Comment by Airblader** [comment]
Our pattern matches usually use regular expressions, so I would prefer to stay with that. However, I would also not say "starts with", but just let the user define the regular expression. Starts with is trivially expressed using the caret anyway.

**Comment by orestisf1993** [comment]
I mixed up the the links. The second one is the one with the regex.

**Comment by HarmtH** [comment]
I like the patch, can this be added? I would also like to propose two additions to this functionality (I could make separate issues for this, but maybe that's a bit premature while it isn't merged yet):
- (Optionally) temporarily unhide the workspace when it's active
- Also skip ignored workspaces with i3-msg workspace (next|prev)(_on_output)? (although this would make it more a general option than an i3bar option, and it could also easily be scripted)

**Comment by clevcode** [comment]
I would love this feature (using a regex for defining workspace names that should not be visible on the i3bar).

My usecase is that I want to only show the workspaces that I'm currently actively using, and to be able to switch between sets of workspaces (i.e. for different projects etc). I often use up to 10 workspaces on each screen, but it gets quite messy, since I have things related to multiple different projects open at the same time, even if I'm trying to use workspaces to separe things as much as I can. Out of sight, out of mind, though.

I'm currently making a small python script, using the i3ipc module, to stash away all the current workspaces (simply by renaming them, adding a chosen prefix, and then creating a new workspace with the original name of the workspace I was at, rather than actually moving the containers on each workspace)

But the "stashed away" workspaces are obviously still visible on the i3bar, which bothers me a bit. It would be great if I could make use of this feature to keep "stashed away" workspaces hidden there.

PS. I was using awesome before switching to i3, and have implemented "awesome style" workspace management using python scripts already, with each monitor having 10 dedicated workspaces (i.e. monitor 1 gets 1-10, monitor 2 gets 11-20, and so on). So rather than binding $mod+1 -> workspace 1 and $mod+Shift+1 -> move container to workspace 1 and so on statically, I invoke my script to switch/move within the workspaces on the current monitor.

**Comment by clevcode** [comment]
(to me personally using a simple prefix to indicate what to hide, such as an underscore, would suffice though)

**Comment by krkhan** [comment]
Similar gist for polybar: https://gist.github.com/krkhan/a6df8684eb8f9db8d1ab25e0168f54b0

**Comment by ymolists** [comment]
@woodensquares thank you so much for providing that gist/patch !!! I have been using your patch for a while to group certain ws in a layer with a local script. Unfortunately this does not apply cleanly to the latest 4.18.1 tag. Would consider revisiting it ? Or at least clarifying the license maybe me or someone else can push that to i3 ? assuming people will accept your proposed changes that only hide certain workspaces that end with a string.

**Comment by orestisfl** [comment]
@ymolists we are not interested in that specific patch, we'd prefer to implement #3818

**Comment by orestisfl** [comment]
My patch from https://github.com/i3/i3/issues/2333#issuecomment-418378118 still applies, maybe you can work with that. I don't remember if it introduced any problems.

**Comment by ymolists** [comment]
@orestisfl i started with that. it has been a long time since i coded in c. the part i reused was the variable names and config parsing which i think was left out as exercise for the user ;-) in your comment. Also the patch has tests which were super helpful for someone new like me. 

I did not want to submit something that looked 90% like the proposed patch and call it my own :-)

**Comment by ymolists** [comment]
The patch also handles unicode comparision [here](https://gist.github.com/woodensquares/c1afc4fb56b4d9d21fa261fb4b28b092#file-ignore-diff-L55) which i was not sure how i would do in c. It was much easier for me to reuse that method.

**Comment by woodensquares** [comment]
@ymolists given there was no interest in going with this patch approach as-is as discussed earlier I did not take this any further, feel free to use it any way you like if it helps.

**Comment by orestisfl** [comment]
@ymolists : I repeat just to make sure you that you don't waste time on this: We do not want to merge a patch for this specific issue, we want to handle the more general case of #3818.

It's good if you want to maintain a temporary patch but you probably don't need to handle configurability and utf8

**Comment by ymolists** [comment]
@orestisfl thank you so much for clarifying that. I will follow the progress on #3818 from now on.

quick question: would it make sense to close this issue then to make it clear for future people coming here from google like i did ?

**Comment by orestisfl** [comment]
Yes, I was thinking about doing it but didn't want to shut off discussion, I'll close it now.
